src/snowflake/connector/connection.py

```diff
@@ -1541,13 +1548,36 @@ def is_an_error(status: QueryStatus) -> bool:
 
     def _all_async_queries_finished(self) -> bool:
         """Checks whether all async queries started by this Connection have finished executing."""
-        queries = copy.copy(
-            self._async_sfqids
-        )  # get_query_status might update _async_sfqids, let's copy the list
-        finished_async_queries = (
-            not self.is_still_running(self.get_query_status(q)) for q in queries
-        )
-        return all(finished_async_queries)
+
+        if not self._async_sfqids:
+            return True
+
+        if sys.version_info >= (3, 8):
+            queries = list(reversed(self._async_sfqids.keys()))
+        else:
+            queries = list(reversed(list(self._async_sfqids.keys())))
+
+        num_workers = min(self.client_prefetch_threads, len(queries))
+        found_unfinished_query = False
+
+        def async_query_check_helper(
+            sfq_id: str,
+        ) -> bool:  # We should upgrade to using cancel_futures=True once supporting 3.9+
+            nonlocal found_unfinished_query
+            return found_unfinished_query or self.is_still_running(
+                self.get_query_status(sfq_id)
+            )
+
+        with ThreadPoolExecutor(
+            max_workers=num_workers, thread_name_prefix="async_query_check_"
+        ) as tpe:
+
+            futures = (tpe.submit(async_query_check_helper, sfqid) for sfqid in queries)
+            for future in as_completed(futures):
+                if not found_unfinished_query and future.result():
```

Comment:
When we support 3.9+ and switch to using [`cancel_futures`](https://docs.python.org/3.9/library/concurrent.futures.html#concurrent.futures.Executor.shutdown), we should handle [`CancelledError`](https://docs.python.org/3.9/library/concurrent.futures.html#concurrent.futures.CancelledError) here.

Reply:
Put this as a comment in the code, otherwise we'll not find it!

Reply:
Sorry, this is an outdated comment.
```
def _all_async_queries_finished():
    tpe = ThreadPoolExecutor()
    futures = (tpe.submit(check_query_status, q) for q in queries)
    for f in as_completed(futures):
        if f.result():
            found_running_query = true
            break
    tpe.shutdown(cancel_futures=true)
    return not found_running_query
```
The code should look something like this, we shouldn't need to access cancelled `future.result()`, after the `break`, so there is no need to handle the exception.